### DISCUSSION TITLE
Extensions: result.GetValueOrDefault() and result.GetValue()

### DISCUSSION THREAD

**Discussion by thebestgin** [louthy/language-ext#1345] [Q&A]
I am looking for an easy way to get the value of a result.
Have I possibly overlooked a possibility or is it missing?

I would like to use it like this:
```c#
SomeEntity someEntity = result.GetValue();
SomeEntity? someNullableEntity = result.GetValueOrDefault();
``` 

Can be implemented very easily via an extension:
```c#
public static class ResultExtensions
{
    public static T? GetValueOrDefault<T>(this Result<T> result)
    {
        return result.Match(
            Succ: entity => entity,
            Fail: _ => default(T)
        );
    }
    
    public static T GetValue<T>(this Result<T> result)
    {
        return result.Match(
            Succ: entity => entity,
            Fail: _ => throw new Exception("Result value null not allowed")
        )!;
    }
}
```

**Comment by louthy** [comment]
`Result` is an intermediate type for `Try` and was never designed for general use.  Instead use `Fin<A>`, it is fully featured and is isomorphic to `Result<A>`.  

> In `v5` of language-ext, which is in beta right now, `Result<A>` has been removed because of this confusion.

If you want to get the value from a `Fin<A>` then you can use:

```c#
Fin<string> ma = FinSucc("thebestgin");

var r1 = ma.Match(Succ: x => $"Hello, {x}", Fail: e => e.ToString());
var r2 = ma.IfFail(e => e.ToString());
var r3 = ma.IfFail("default value");
```
If you definitely know the case, then you can cast:
```c#
if(ma)  // if in a `Succ` state
{
    r4 = (string)ma;
}

if(!ma)  // if in a `Fail` state
{
    r5 = (Error)ma;
}
```
In language-ext `v5` you can also use C#'s pattern-matching:
```c#
var r6 = ma switch
{
    Fin.Succ<string>(var x) => x,
    Fin.Fail<string>(var e) => e.ToString()
}
```

**Reply by thebestgin** [reply]
Hi Louthy,

Thank you very much for pointing this out to me. 
I'll use the Fin type from now on.

Very cool code snippet. I was really pleased.
Exactly what I was looking for.
